### PR TITLE
[6.x] Add requeue function to jobs

### PR DESCRIPTION
**Pull request by andershagbard**
This pull requests adds a `requeue` method to database and Redis jobs. The idea comes from https://github.com/laravel/ideas/issues/1381

The method does the same as the `release` method, but before putting the job back into queue, it decrements the attempts attribute.

This handles an issue like this, where the job could fail, because the attempts attribute had exceeded the tries attribute.

```php
Redis::throttle('key')->allow(10)->every(60)->then(function () {
    try {
        // API logic
    }
    catch (Exception) {
        // Release the job back into the queue
        return $this->release(10);
    }
}, function () {
    // Could not obtain lock...
    return $this->requeue(10);
});
```

### CODE REVIEW

**Comment by sisve** [comment]
This is what the release method did back in 5.2.40, until there was a [change to the markJobAsReserved](https://github.com/laravel/framework/commit/c7975cb00865178c6c55cb875788314fa20cc1e8#diff-b05083dc38b4e45d38d28c676abbad83R216) method. If we change markJobAsReserved into only marking the job as reserved (as the name suggests), then the release method will work as intended.

Does this requeueing work differently than if the job where to dispatch itself with the DispatchesJob traits and `$this->dispatch($this)`?

**Comment by andershagbard** [comment]
> This is what the release method did back in 5.2.40, until there was a [change to the markJobAsReserved](https://github.com/laravel/framework/commit/c7975cb00865178c6c55cb875788314fa20cc1e8#diff-b05083dc38b4e45d38d28c676abbad83R216) method. If we change markJobAsReserved into only marking the job as reserved (as the name suggests), then the release method will work as intended.

It's important to have both methods available. They both have use cases, as in my initial example.

> Does this requeueing work differently than if the job where to dispatch itself with the DispatchesJob traits and `$this->dispatch($this)`?

Works exactly the same way as the `release` method, just decrements the `attempts` attribute before sending back to queue.

**Comment by taylorotwell** [comment]
Can we discuss what advantage this provides over re-queueing a new instance of the job?

File: src/Illuminate/Queue/Jobs/DatabaseJob.php [hunk 1]
```diff
@@ -56,6 +56,23 @@ public function release($delay = 0)
...
 
+    /**
+     * Requeue the job.
+     *
+     * @param  int   $delay
+     * @return void
+     */
+    public function requeue($delay = 0)
+    {
+        parent::release($delay);
```
Comment (taylorotwell) [inline src/Illuminate/Queue/Jobs/DatabaseJob.php]: Why do you call this here?

File: src/Illuminate/Queue/Jobs/RedisJob.php [hunk 2]
```diff
@@ -97,6 +97,19 @@ public function release($delay = 0)
...
 
+    /**
+     * Requeue the job.
+     *
+     * @param  int   $delay
+     * @return void
+     */
+    public function requeue($delay = 0)
+    {
+        parent::release($delay);
```
Comment (taylorotwell) [inline src/Illuminate/Queue/Jobs/RedisJob.php]: Why do you call this here?

**Comment by taylorotwell** [comment]
Has no real tests.

**Comment by andershagbard** [comment]
I've made some changes. _Not sure why the don't show up on this PR?_

https://github.com/andershagbard/framework/commit/226f25545cff5444ad9152b2c384368a9a7f2764
https://github.com/andershagbard/framework/commit/92c51a80578c432fbafdc34480b2396c9a02e3f8

These should add properly tests for Redis and Database jobs.

One thing I think might be "missing" is a method like this one:

```php
public function isDeletedOrReleased()
{
    return $this->isDeleted() || $this->isReleased();
}
```

But naming it `isDeletedReleasedOrRequeued` feels very weird.

**Comment by GrahamCampbell** [comment]
If your new code has tests, please feel free to open a new PR. Your commits will show up then. :)

**Pull request closed without merging**